## Week 11, Lab 1: SQL Practice

> Author: Matt Brems

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the famous (but fake) Northwind database. It contains various tables of transactions. The schema for the Northwind database can be found [here](http://www.zentut.com/wp-content/uploads/downloads/2013/06/Northwind-Sample-Database-Diagram.pdf).

In [1]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgres://postgres:iloveindia@localhost:5432/northwind')

### Prompt 1

How many rows are there in the `customers` table of the Northwind database?

In [2]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*) 
FROM customers;
"""

pd.read_sql_query(sql, engine)

count
0     91

**Prompt 1 Answer**: 91

### Prompt 2

If you sort the customers in alphabetical order by name, which customer is fourth?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT * 
FROM customers
ORDER BY customers."ContactName";
"""

pd.read_sql_query(sql, engine)

CustomerID                           CompanyName              ContactName  \
0       ROMEY                      Romero y tomillo         Alejandra Camino   
1       MORGK                Morgenstern Gesundkost          Alexander Feuer   
2       ANATR    Ana Trujillo Emparedados y helados             Ana Trujillo   
3       TRADH                Tradição Hipermercados        Anabela Domingues   
4       GOURL                   Gourmet Lanchonetes            André Fonseca   
5       EASTC                    Eastern Connection                Ann Devon   
6       LAMAI                      La maison d'Asie           Annette Roulet   
7       ANTON               Antonio Moreno Taquería           Antonio Moreno   
8       FAMIA                    Familia Arquibaldo                Aria Cruz   
9       SPLIR                 Split Rail Beer & Ale       Art Braunschweiger   
10      QUEDE                           Que Delícia         Bernardo Batista   
11      FRANR                   France restauration           Carine Schmitt   
12      LILAS                     LILA-Supermercado          Carlos González   
13      HILAA                      HILARION-Abastos         Carlos Hernández   
14      MAISD                          Maison Dewey          Catherine Dewey   
15      BERGS                    Berglunds snabbköp       Christina Berglund   
16      LACOR                  La corne d'abondance            Daniel Tonini   
17      FISSA  FISSA Fabrica Inter. Salchichas S.A.               Diego Roel   
18      SPECD                  Spécialités du monde        Dominique Perrier   
19      GALED                Galería del gastrónomo         Eduardo Saavedra   
20      CONSH                 Consolidated Holdings          Elizabeth Brown   
21      BOTTM                 Bottom-Dollar Markets        Elizabeth Lincoln   
22      LINOD                      LINO-Delicateses         Felipe Izquierdo   
23      LONEP              Lonesome Pine Restaurant              Fran Wilson   
24      CENTC            Centro comercial Moctezuma          Francisco Chang   
25      BLONP              Blondesddsl père et fils       Frédérique Citeaux   
26      PICCO                      Piccolo und mehr              Georg Pipps   
27      MAGAA          Magazzini Alimentari Riuniti         Giovanni Rovelli   
28      PERIC             Pericles Comidas clásicas      Guillermo Fernández   
29      BLAUS               Blauer See Delikatessen               Hanna Moos   
..        ...                                   ...                      ...   
61      WILMK                           Wilman Kala          Matti Karttunen   
62      REGGC                    Reggiani Caseifici          Maurizio Moroni   
63      RICSU                    Richter Supermarkt             Michael Holz   
64      TORTU                   Tortuga Restaurante     Miguel Angel Paolino   
65      VAFFE                          Vaffeljernet              Palle Ibsen   
66      FRANS                        Franchi S.p.A.            Paolo Accorti   
67      SUPRD                      Suprêmes délices         Pascale Cartrain   
68      HUNGO          Hungry Owl All-Night Grocers         Patricia McKenna   
69      CACTU            Cactus Comidas para llevar         Patricio Simpson   
70      VINET             Vins et alcools Chevalier             Paul Henriot   
71      WELLI                Wellington Importadora            Paula Parente   
72      RATTC            Rattlesnake Canyon Grocery             Paula Wilson   
73      COMMI                      Comércio Mineiro             Pedro Afonso   
74      FRANK                        Frankenversand            Peter Franken   
75      KOENE                       Königlich Essen            Philip Cramer   
76      WARTH                        Wartian Herkku         Pirkko Koskitalo   
77      LEHMS                   Lehmanns Marktstand           Renate Messner   
78      OLDWO                Old World Delicatessen            Rene Phillips   
79 

**Prompt 2 Answer**: Ana Trujillo.

### Prompt 3

Among those customers who live in Germany, if you sort the customers in alphabetical order by name, which customer is fourth?

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT * 
FROM customers
WHERE customers."Country" = 'Germany'
ORDER BY customers."ContactName";
"""

pd.read_sql_query(sql, engine)

CustomerID               CompanyName          ContactName  \
0       MORGK    Morgenstern Gesundkost      Alexander Feuer   
1       BLAUS   Blauer See Delikatessen           Hanna Moos   
2       OTTIK        Ottilies Käseladen  Henriette Pfalzheim   
3       QUICK                QUICK-Stop          Horst Kloss   
4       TOMSP        Toms Spezialitäten        Karin Josephs   
5       ALFKI       Alfreds Futterkiste         Maria Anders   
6       FRANK            Frankenversand        Peter Franken   
7       KOENE           Königlich Essen        Philip Cramer   
8       LEHMS       Lehmanns Marktstand       Renate Messner   
9       WANDK         Die Wandernde Kuh          Rita Müller   
10      DRACD  Drachenblut Delikatessen         Sven Ottlieb   

            ContactTitle             Address            City Region  \
0    Marketing Assistant         Heerstr. 22         Leipzig   None   
1   Sales Representative      Forsterstr. 57        Mannheim   None   
2                  Owner  Mehrheimerstr. 369            Köln   None   
3     Accounting Manager    Taucherstraße 10       Cunewalde   None   
4      Marketing Manager       Luisenstr. 48         Münster   None   
5   Sales Representative       Obere Str. 57          Berlin   None   
6      Marketing Manager   Berliner Platz 43         München   None   
7        Sales Associate       Maubelstr. 90     Brandenburg   None   
8   Sales Representative        Magazinweg 7  Frankfurt a.M.   None   
9   Sales Representative   Adenauerallee 900       Stuttgart   None   
10   Order Administrator        Walserweg 21          Aachen   None   

   PostalCode  Country         Phone           Fax  
0       04179  Germany   0342-023176          None  
1       68306  Germany    0621-08460    0621-08924  
2       50739  Germany  0221-0644327  0221-0765721  
3       01307  Germany   0372-035188          None  
4       44087  Germany   0251-031259   0251-035695  
5       12209  Germany   030-0074321   030-0076545  
6       80805  Germany   089-0877310   089-0877451  
7       14776  Germany    0555-09876          None  
8       60528  Germany   069-0245984   069-0245874  
9       70563  Germany   0711-020361   0711-035428  
10      52066  Germany   0241-039123   0241-059428

**Prompt 3 Answer**: Horst Kloss.

### Prompt 4

From the `employees` table, what are the first names of the individuals who have a birthdate in the 1960s?

In [9]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT employees."FirstName"
FROM employees
WHERE employees."BirthDate" BETWEEN '1-1-1960' AND '12-31-1969';
"""

pd.read_sql_query(sql, engine)

FirstName
0     Janet
1   Michael
2    Robert
3      Anne

**Prompt 4 Answer**: There are four employees born in the 1960s: Janet, Michael, Robert, and Anne.

### Prompt 5

In `order_details` table of the Northwind database, what is the total number of units sold for products 2, 3, and 4? 
> You should give us three numbers - one for each `ProductID`.

In [10]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT order_details."ProductID", SUM(order_details."Quantity") 
FROM order_details
WHERE order_details."ProductID" BETWEEN 2 AND 4
GROUP BY order_details."ProductID"
ORDER BY order_details."ProductID";
"""

pd.read_sql_query(sql, engine)

ProductID   sum
0          2  1057
1          3   328
2          4   453

**Prompt 5 Answer**: 
- 1,057 units of ProductID 2 have been sold, 
- 328 units of ProductID 3 have been sold, and 
- 453 units of ProductID 4 have been sold.

### Prompt 6

From the `order_details` table of the Northwind database, calculate the revenue per product. In your SQL output, rename the revenue column `Revenue` and sort your results from largest revenue per product to smallest revenue per product. Which `ProductID` has the fifth-highest revenue, and what is that revenue rounded to the nearest dollar? 
> Note that revenue for a given product should be equal to the total unit price of that product times the quantity. You do not need to worry about the discount here. If you haven't multiplied columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [11]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT order_details."ProductID", SUM(order_details."Quantity" * order_details."UnitPrice") AS "Revenue"
FROM order_details
GROUP BY order_details."ProductID"
ORDER BY "Revenue" DESC
LIMIT 5;
"""

pd.read_sql_query(sql, engine)

ProductID        Revenue
0         38  149984.200821
1         29   87736.400513
2         59   76296.000000
3         60   50286.000374
4         62   49827.899998

**Prompt 6 Answer**: ProductID 62 has the fifth-highest revenue, at \$49,828.

### Prompt 7

From the `products` table of the Northwind database, show the `ProductName`, `UnitsInStock`, and `UnitsOnOrder` values for those products who have more units on order than units in stock. Among these results, which product has the most units on order?

In [12]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT products."ProductName", products."UnitsInStock", products."UnitsOnOrder"
FROM products
WHERE products."UnitsInStock" < products."UnitsOnOrder";
"""

pd.read_sql_query(sql, engine)

ProductName  UnitsInStock  UnitsOnOrder
0                       Chang            17            40
1               Aniseed Syrup            13            70
2              Queso Cabrales            22            30
3         Sir Rodney's Scones             3            40
4           Gorgonzola Telino             0            70
5          Mascarpone Fabioli             9            40
6                  Gravad lax            11            50
7                 Rogede sild             5            70
8                   Chocolade            15            70
9                    Maxilaku            10            60
10  Wimmers gute Semmelknödel            22            80
11  Louisiana Hot Spiced Okra             4           100
12        Scottish Longbreads             6            10
13              Longlife Tofu             4            20

**Prompt 7 Answer**: Louisiana Hot Spiced Okra has the most units on order, with 100.

### Prompt 8

From the `order_details` table, show the products that have an average discount of at least 8%. Among these results, which average discount is closest to 8%?

In [9]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT order_details."ProductID", AVG(order_details."Discount")
FROM order_details
GROUP BY order_details."ProductID"
HAVING AVG(order_details."Discount") >= 0.08
ORDER BY AVG(order_details."Discount") DESC;
"""

pd.read_sql_query(sql, engine)

ProductID       avg
0         48  0.108333
1          2  0.102273
2          9  0.100000
3         58  0.100000
4         25  0.086111
5         30  0.081250

**Prompt 8 Answer**: Among all ProductIDs with an average discount of at least 8%, ProductID 30 has an average discount closest to 8% with an average discount of 8.125%.

### Prompt 9

Return the `orders` table, but also include the first name, last name, and country of the employee assigned to each order. What is the first and last name of the employee assigned to `OrderID` 11077? (This is the last row in the data, provided that you do not sort.)

In [10]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT orders.*, employees."FirstName", employees."LastName", employees."Country"
FROM orders
LEFT JOIN employees
ON orders."EmployeeID" = employees."EmployeeID";
"""

pd.read_sql_query(sql, engine)

OrderID CustomerID  EmployeeID   OrderDate RequiredDate ShippedDate  \
0      10248      VINET           5  1996-07-04   1996-08-01  1996-07-16   
1      10249      TOMSP           6  1996-07-05   1996-08-16  1996-07-10   
2      10250      HANAR           4  1996-07-08   1996-08-05  1996-07-12   
3      10251      VICTE           3  1996-07-08   1996-08-05  1996-07-15   
4      10252      SUPRD           4  1996-07-09   1996-08-06  1996-07-11   
5      10253      HANAR           3  1996-07-10   1996-07-24  1996-07-16   
6      10254      CHOPS           5  1996-07-11   1996-08-08  1996-07-23   
7      10255      RICSU           9  1996-07-12   1996-08-09  1996-07-15   
8      10256      WELLI           3  1996-07-15   1996-08-12  1996-07-17   
9      10257      HILAA           4  1996-07-16   1996-08-13  1996-07-22   
10     10258      ERNSH           1  1996-07-17   1996-08-14  1996-07-23   
11     10259      CENTC           4  1996-07-18   1996-08-15  1996-07-25   
12     10260      OTTIK           4  1996-07-19   1996-08-16  1996-07-29   
13     10261      QUEDE           4  1996-07-19   1996-08-16  1996-07-30   
14     10262      RATTC           8  1996-07-22   1996-08-19  1996-07-25   
15     10263      ERNSH           9  1996-07-23   1996-08-20  1996-07-31   
16     10264      FOLKO           6  1996-07-24   1996-08-21  1996-08-23   
17     10265      BLONP           2  1996-07-25   1996-08-22  1996-08-12   
18     10266      WARTH           3  1996-07-26   1996-09-06  1996-07-31   
19     10267      FRANK           4  1996-07-29   1996-08-26  1996-08-06   
20     10268      GROSR           8  1996-07-30   1996-08-27  1996-08-02   
21     10269      WHITC           5  1996-07-31   1996-08-14  1996-08-09   
22     10270      WARTH           1  1996-08-01   1996-08-29  1996-08-02   
23     10271      SPLIR           6  1996-08-01   1996-08-29  1996-08-30   
24     10272      RATTC           6  1996-08-02   1996-08-30  1996-08-06   
25     10273      QUICK           3  1996-08-05   1996-09-02  1996-08-12   
26     10274      VINET           6  1996-08-06   1996-09-03  1996-08-16   
27     10275      MAGAA           1  1996-08-07   1996-09-04  1996-08-09   
28     10276      TORTU           8  1996-08-08   1996-08-22  1996-08-14   
29     10277      MORGK           2  1996-08-09   1996-09-06  1996-08-13   
..       ...        ...         ...         ...          ...         ...   
800    11048      BOTTM           7  1998-04-24   1998-05-22  1998-04-30   
801    11049      GOURL           3  1998-04-24   1998-05-22  1998-05-04   
802    11050      FOLKO           8  1998-04-27   1998-05-25  1998-05-05   
803    11051      LAMAI           7  1998-04-27   1998-05-25        None   
804    11052      HANAR           3  1998-04-27   1998-05-25  1998-05-01   
805    11053      PICCO           2  1998-04-27   1998-05-25  1998-04-29   
806    11054      CACTU           8  1998-04-28   1998-05-26        None   
807    11055      HILAA           7  1998-04-28   1998-05-26  1998-05-05   
808    11056      EASTC           8  1998-04-28   1998-05-12  1998-05-01   
809    11057      NORTS           3  1998-04-29   1998-05-27  1998-05-01   
810    11058      BLAUS           9  1998-04-29   1998-05-27        None   
811    11059      RICAR           2  1998-04-29   1998-06-10        None   
812    11060      FRANS           2  1998-04-30   1998-05-28  1998-05-04   
813    11061      GREAL           4  1998-04-30   1998-06-11        None   
814    11062      REGGC           4  1998-04-30   1998-05-28        None   
815    11063      HUNGO           3  1998-04-30   1998-05-28  1998-05-06   
816    11064      SAVEA           1  1998-05-01   1998-05-29  1998-05-04   
817    11065      LILAS           8  1998-05-01   1998-05-29        None   
818    11066      WHITC           7  1998-05-01   1998-05-29  1998-05-04   
819    11067      DRACD           1  1998-05-04   1998-05-18  1998-05-06   
820    11068      QUEEN           8  1998-05-04   199

**Prompt 9 Answer**: `OrderID` 11077 corresponds to Nancy Davolio.

### Prompt 10

Find the average amount of freight across all orders, broken out by country of *employee*. (**NOT `ShipCountry`**.) How much higher is the average freight from UK employees than the average freight from US employees?

> For a bonus, round the freight to the nearest whole number and rename the column `Average Freight`.

In [11]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT employees."Country", AVG(orders."Freight") AS "Average Freight"
FROM orders
LEFT JOIN employees
ON orders."EmployeeID" = employees."EmployeeID"
GROUP BY employees."Country";
"""

pd.read_sql_query(sql, engine)

Country  Average Freight
0      UK        78.977143
1     USA        77.973284

**Prompt 10 Answer**: The average freight for orders shipped by UK employees is about 1 unit higher than the average freight shipped by US employees.